### ISSUE TITLE
derive_partial_eq_without_eq: False positive for generic struct

### ISSUE THREAD

**Issue by Turbo87** [rust-lang/rust-clippy#9868]
### Summary

see "Reproducer" section below :)

### Lint Name

derive_partial_eq_without_eq

### Reproducer

I tried this code:

```rust
use num_traits::Float;

#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
pub struct FlatProjection<T: Float> {
    kx: T,
    ky: T,

    lat: T,
    lon: T,
}
```

(see https://github.com/Turbo87/flat-projection-rs/blob/8490b90e0f4ecff5ba47f41803007e6dc9ae455c/src/lib.rs#L80-L87)

I saw this happen:

```
warning: you are deriving `PartialEq` and can implement `Eq`
  --> src/lib.rs:80:30
   |
80 | #[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
   |                              ^^^^^^^^^ help: consider deriving `Eq` as well: `PartialEq, Eq`
   |
   = note: `#[warn(clippy::derive_partial_eq_without_eq)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#derive_partial_eq_without_eq
```

I expected to see this happen:

If I understand correctly, we can only derive `Eq` if all fields of the struct do so too. But the `Float` trait does not implement it as far as I understand, so the rule shouldn't trigger.

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-apple-darwin
release: 1.65.0
LLVM version: 15.0.0
```

**Comment by Alexendoo** [comment]
The derived impls have a bound on any generics that appear in the type, in this example it would be roughly

```rust
impl<T: Eq + Float> Eq for FlatProjection<T> {}
```

So the struct will be `Eq` only if `T` is, [`OrderedFloat`](https://docs.rs/ordered-float/latest/ordered_float/struct.OrderedFloat.html) would be the notable example of something that is both `Eq` and `Float`

**Comment by Turbo87** [comment]
huh, interesting. thanks for the explanation!
